lib/winapi/filemanip.c

```diff
@@ -235,3 +239,80 @@ BOOL MoveFileA (LPCTSTR lpExistingFileName, LPCTSTR lpNewFileName)
         return TRUE;
     }
 }
+
+DWORD GetLogicalDrives (VOID)
+{
+    DWORD result = 0;
+    NTSTATUS status;
+    ANSI_STRING path;
+    HANDLE handle;
+    OBJECT_ATTRIBUTES attributes;
+    struct {
+        OBJECT_DIRECTORY_INFORMATION objDirInfo;
+        CHAR filenameBuf[100];
+    } objDirInfoBuf;
+
+    InitializeObjectAttributes(&attributes, NULL, OBJ_CASE_INSENSITIVE, ObDosDevicesDirectory(), NULL);
+    status = NtOpenDirectoryObject(&handle, &attributes);
+
+    if (!NT_SUCCESS(status)) {
+        SetLastError(RtlNtStatusToDosError(status));
+        return 0;
+    }
+
+    for (ULONG index = 0;;) {
+        POBJECT_DIRECTORY_INFORMATION objDirInfo = (POBJECT_DIRECTORY_INFORMATION)&objDirInfoBuf;
+        status = NtQueryDirectoryObject(handle, objDirInfo, sizeof(objDirInfoBuf), (index == 0), &index, NULL);
+
+        if (status == STATUS_NO_MORE_ENTRIES) {
+            break;
+        } else if (!NT_SUCCESS(status)) {
+            NtClose(handle);
+            SetLastError(RtlNtStatusToDosError(status));
+            return 0;
+        }
+
+        // Ignore entries like "CdRom0:"
+        if (objDirInfo->Name.Length != 2 || objDirInfo->Name.Buffer[1] != ':') {
+            continue;
+        }
+
+        if (isalpha(objDirInfo->Name.Buffer[0])) {
+            result |= 1 << (toupper(objDirInfo->Name.Buffer[0]) - 'A');
+        }
+    }
+
+    NtClose(handle);
+    SetLastError(ERROR_SUCCESS);
+    return result;
+}
+
+DWORD GetLogicalDriveStringsA (DWORD nBufferLength, LPSTR lpBuffer)
+{
+    DWORD driveBitmask = GetLogicalDrives();
+    if (driveBitmask == 0 && GetLastError() != ERROR_SUCCESS) {
+        return 0;
+    }
+
+    DWORD requiredBufLength = 0;
+    for (int bit = 0; bit < 26; bit++) {
+        if (driveBitmask & (1 << bit)) {
+            requiredBufLength += 4;
+        }
+    }
+
+    if (nBufferLength == 0 || nBufferLength < requiredBufLength) {
+        return requiredBufLength+1;
+    }
+
+    for (int bit = 0; bit < 26; bit++) {
+        if (driveBitmask & (1 << bit)) {
+            *lpBuffer++ = 'A' + bit;
+            *lpBuffer++ = ':';
+            *lpBuffer++ = '\\';
```

Comment:
The MSDN isn't clear about the format of the returned strings, did you confirm this format on Windows?

Reply:
Yes.